[DEVI-1570] Updates PagerDuty App docs for Scoped OAuth (#68)
PagerDuty/developer-docs: docs/app-integration-development/19-OAuth-Functionality.md

---
tags: [app-integration-development]
---

# OAuth Functionality

## What is OAuth functionality?

PagerDuty Apps can use OAuth functionality to obtain access tokens that allow the application to interact with the PagerDuty [REST API](/api-reference/) on behalf of a user or as the app itself.

For example, an application with its own UI may wish to call the PagerDuty REST API _as a user_ to obtain data specific to that user. An application that receives PagerDuty webhooks or performs other server-side processing may wish to call back to the PagerDuty _as the app_ to obtain more information or update state.

The capabilities and behavior differ based on whether you choose to use the Scoped OAuth or Classic User OAuth functionality.

> ## OAuth functionality or PagerDuty API key?

Reviewer: ```suggestion
## OAuth or PagerDuty API key?
```

> A PagerDuty App with OAuth functionality is the preferred choice for both third-party integrations and long-lived applications that interact with a single PagerDuty Account.

Reviewer: It's a bit awkward because some OAuth clients can interact with multiple accounts, so it's best not to mention the number of accounts at all?

```suggestion
A PagerDuty App with OAuth functionality is the preferred choice for both third-party integrations and long-lived applications.
```

Reviewer: > Are we talking about that fact that Classic User OAuth can interact with multiple PagerDuty Accounts even when it is not published?

I mean more so, classic user oauth clients _can_ be published right now and interact with multiple PagerDuty accounts.

User API Keys are useful for scripts and personal projects but are also tied to that single user. If the user leaves the account then the key is disabled and the application will no longer function properly. Account API Keys have full access to a PagerDuty Account.

## Types of OAuth functionality

### Classic User OAuth

With Classic User OAuth the application is always acting as a PagerDuty User. The application must take the user through an OAuth 2.0 code flow to obtain their authorization and consent before being granted an act as user access token. See [Act as User Tokens via Code Grant](09-Act-as-User-OAuth-Code-Grant.md) and [Act as User Tokens via PKCE](10-Act-as-User-OAuth-PKCE.md) for the two methods of obtaining these tokens. See below for a discussion of the differences between these two methods.

> The access available to an application using Classic User OAuth is the intersection of the scopes granted to the application and the permissions of the user the application is acting on behalf of. Scopes in Classic User OAuth are limited to either `read` which allows read-only access to all resources or `write` which allows read/write access to all resources.

Reviewer: ```suggestion
The access available to an application using Classic User OAuth is the intersection of the scopes granted to the application and the permissions of the user the application is acting on behalf of. Scopes in Classic User OAuth are limited to either `read` which allows read-only access to all resources available to the authorizing user, or `write` which allows read/write access to all resources available to the authorizing user.
```

> Classic User OAuth works for both private and public PagerDuty Apps.

Reviewer: ```suggestion
Classic User OAuth clients can also be published for use with other PagerDuty accounts.
```

Scoped OAuth allows an application to act to outside the context of a user by obtaining an act as app access token using an OAuth 2.0 client credentials flow. Optionally, it can also obtain act as user tokens after obtaining their authorization and consent. See [Act as App Tokens](20-Act-as-App-Token.md) and [Act as User Tokens via Code Grant](09-Act-as-User-OAuth-Code-Grant.md) for the details of obtaining each type of token.

The access available to an application using Scoped OAuth differs by token type:
* When using an act as app token, access is only limited by the scopes granted to the application. If the application has the `incidents.read` scope, it will be able to read all incidents on the account.
* When using an act as user token, access is the intersection of the scopes granted to the application and the permissions of the user. If an application has the `incidents.read` scope and is acting as user Pagey, it will only be able to read incidents that user Pagey has permission to see.

If a [REST API](/api-reference/) endpoint works with Scoped OAuth, the documentation for that endpoint will say "Scoped OAuth requires:" and list the required scope to access the endpoint.

> Scoped OAuth only works for private PagerDuty Apps at this time. The PagerDuty App is limited to accessing the PagerDuty Account that created it.

Reviewer: ```suggestion
Scoped OAuth clients cannot currently be published to allow use with other PagerDuty accounts -- the PagerDuty App is limited to accessing the PagerDuty Account that created it.
```

### Understanding what type of OAuth functionality to use

#### Scoped OAuth offers improved security
Scoped OAuth improves on Classic User OAuth by offering a set of scopes that individually grant access to read or write a single type of resource. For example, the `incidents.read` scope is only able to read incidents on an account. The `incidents.write` scope is able to create, update, or delete an incident but is not able to read existing incidents.

#### Scoped OAuth can act as an app or as a user

> Scoped OAuth has the ability to outside the context of a user as a PagerDuty App. This is useful in server to server flows where a user may not be involved. It also eliminates the need to manage service or bot user accounts. When configured with a redirect URI, Scoped OAuth applications can also obtain act as user tokens with the [Act as User Tokens via Code Grant](09-Act-as-User-OAuth-Code-Grant.md) flow to obtain a user's authorization and consent.

Reviewer: ```suggestion
Scoped OAuth has the ability to act outside the context of a user, as a PagerDuty App. This is useful in server-to-server flows, where a user may not be involved. It also eliminates the need to manage service or bot user accounts. When configured with a redirect URI, Scoped OAuth applications can also use [Authorization Code Grant](09-Act-as-User-OAuth-Code-Grant.md) to act as a user, with that user's authorization and consent.
```

> #### Classic User OAuth is still the only supported mechanism for public apps

Reviewer: ```suggestion
#### Only Classic User OAuth can be used with other PagerDuty accounts
```

> If you are building an integration on top of PagerDuty with the intention of making it usable across many PagerDuty Accounts then Classic User OAuth continues to be the right choice.

Reviewer: ```suggestion
If you are building an integration on top of PagerDuty with the intention of publishing it for use with many PagerDuty Accounts, then Classic User OAuth continues to be the right choice.
```

#### Scoped OAuth applications should be able to securely store the client secret
All Scoped OAuth applications have a `client_id` and `client_secret` that can be used to obtain act as app tokens. Applications using Scoped OAuth functionality should have a server-side component where the `client_secret` is properly secured.

#### Code Grant or PKCE?

> Scoped OAuth functionality is only able to obtain [Act as User Tokens via Code Grant](09-Act-as-User-OAuth-Code-Grant.md) at this time.

Reviewer: I know this is about auth code vs auth code with PKCE, but it feels weird not to mention client credentials grant here.

Author: Yeah I didn't do that here because this primarily feels about "PKCE versus Code Grant" and the whole "public / confidential client" thing. I'm open to a separate "which flow should I use" question later...

Reviewer: PKCE is still the Authorization Code flow though .. maybe it should be "To PKCE or not to PKCE?"

Author: Updated.
